theano/tensor/nnet/bn.py

```diff
@@ -0,0 +1,57 @@
+import theano
+from theano.scalar import Composite
+from theano.scalar import add, sub, true_div, mul
+
+
+class BNComposite(Composite):
+
+    def __init__(self, dtype):
+        x = theano.scalar.Scalar(dtype=dtype).make_variable()
+        mean = theano.scalar.Scalar(dtype=dtype).make_variable()
+        std = theano.scalar.Scalar(dtype=dtype).make_variable()
+        gamma = theano.scalar.Scalar(dtype=dtype).make_variable()
+        beta = theano.scalar.Scalar(dtype=dtype).make_variable()
+        o = add(mul(true_div(sub(x, mean), std), gamma), beta)
+        inputs = [x, mean, std, gamma, beta]
+        outputs = [o]
+        super(BNComposite, self).__init__(inputs, outputs)
+
+    def grad(self, inps, grads):
+        x, mean, std, gamma, beta = inps
+        top, = grads
+        dx = (top * gamma) / std
+        dmean = -(top * gamma) / std
+        dstd = -(top * gamma * (x - mean)) / (std * std)
+        dgamma = top * (x - mean) / std
+        return [dx, dmean, dstd, dgamma, top]
+
+
+def batch_normalization(inputs, gamma, beta, mean, std):
+    """
+    This function will build the symbolic graph for applying batch normalization
+    to a set of activations. As no intermediate representations are stored for the
+    back-propagation, this implementation lower the memory usage, however,
+    it is 5-10% slower than a naive theano implementation, as it redo
+    some foward computations for the backprop.
```

Comment:
Then maybe there should be a `mode` argument or something that defaults to `"low_mem"` (where it uses this Op) but can also be set to `"fast"` (where it just directly returns `((inputs - mean) / std) * gamma + beta`)?

@nouiz: What about the optimizations you talked about in https://github.com/Lasagne/Lasagne/issues/141#issuecomment-129663016 and following -- is there a way to make use of them here in future? Or will they be used automatically when they are implemented along with an optimizer? I guess if there's a `batch_normalization()` function in Theano, people will expect it to be faster/better than writing down the expression themselves.